### PR TITLE
CRL generation master: new utility to enable|disable

### PR DESCRIPTION
**Pull request by flo-renaud**
### CRL generation master: new utility to enable|disable
Implement a new command ipa-clrgen-manage to enable, disable, or check the status of CRL generation on the localhost.
The command automates the manual steps described in the wiki https://www.freeipa.org/page/Howto/Promote_CA_to_Renewal_and_CRL_Master

Fixes: https://pagure.io/freeipa/issue/5803

### Test: add new tests for ipa-crlgen-manage
Add new integration tests for the new command ipa-crlgen-manage, and test_cmdline tests.

Related to: https://pagure.io/freeipa/issue/5803

### CODE REVIEW

**Comment by flo-renaud** [comment]
This PR is still Work In Progress:

- [ ] I need to figure out how to force the generation of CRL when `enable` is called.
- [ ] the [design](https://www.freeipa.org/page/V4/Promotion_to_CRL_generation_master) page hasn't been updated yet and still points to the previous try with `ipa-advise`.

But feel free to start reviewing and send comments.

File: ipaserver/install/cainstance.py [hunk 1]
```diff
@@ -1386,6 +1390,149 @@ def __dogtag10_migration(self):
...
+        logger.debug("%s successfully started", self.service_name)
+
+        # TBD: make sure a CRL is generated if setup_crl is True
+
+        # Edit the RewriteRule
+        def comment_rewriterule():
+            logger.info("Editing %s", paths.HTTPD_IPA_PKI_PROXY_CONF)
+            # look for the pattern RewriteRule ^/ipa/crl/MasterCRL.bin ..
+            # and comment out
+            pattern = r"^\s*(RewriteRule\s+\^/ipa/crl/MasterCRL.bin\s.*)$"
```
Comment (rcritten) [inline ipaserver/install/cainstance.py]: I wonder if we should either make this value a constant and/or make sure we mention it somewhere in the proxy config file so if it changes in one place it is also changed in the other.

**Comment by rcritten** [comment]
Looking pretty good to me, moving in the right direction.

I don't know what to do about the Apache config. It just seems like something years from now we'd make some minor change and break things. I suppose the unit test would catch it...

**Comment by flo-renaud** [comment]
PR updated with:
- new code to trigger CRL update when ipa-crlgen-manage enable is called
- new code to prevent server uninstallation if it is CRL generation master, unless --ignore-last-of-role is specified (or in interactive mode, if the user positively answers to the prompt).
- test for uninstall

**Comment by netoarmando** [comment]
@flo-renaud could you please force-push again to try to trigger PR-CI?

**Comment by fcami** [comment]
Hi @flo-renaud could you please rebase on top of https://github.com/freeipa/freeipa/commit/8327e11b6be6c675a2eb7b453a5e7fb68d0c12c2 ?

**Comment by flo-renaud** [comment]
@fcami, thanks for the hint, PR rebased
@netoarmando it seems PRCI is disturbed by a force-push if the last commit did not change (I added a new before-last commit and the last commit hash did not change, it's probably why PRCI wasn't re-triggered).

**Comment by fcami** [comment]
On master server:
[root@f29-idm0 ~]# ipa-crlgen-manage status
CRL generation: enabled
Unable to find last CRL
The ipa-crlgen-manage command was successful
[root@f29-idm0 ~]# ipa-crlgen-manage disable
Stopping pki-tomcatd
Editing /var/lib/pki/pki-tomcat/conf/ca/CS.cfg
Starting pki-tomcatd
Editing /etc/httpd/conf.d/ipa-pki-proxy.conf
Restarting httpd
CRL generation disabled on the local host. Please make sure to configure CRL generation on another master with ipa-crlgen-manage enable
The ipa-crlgen-manage command was successful

On replica:
[root@f29-idm1 ~]# ipa-crlgen-manage status

CRL generation: disabled
The ipa-crlgen-manage command was successful
[root@f29-idm1 ~]# 
[root@f29-idm1 ~]# ipa-crlgen-manage disable
Nothing to do, CRL generation already disabled
CRL generation disabled on the local host. Please make sure to configure CRL generation on another master with ipa-crlgen-manage enable
The ipa-crlgen-manage command was successful
[root@f29-idm1 ~]# ipa-crlgen-manage enable
Stopping pki-tomcatd
Editing /var/lib/pki/pki-tomcat/conf/ca/CS.cfg
Starting pki-tomcatd
Editing /etc/httpd/conf.d/ipa-pki-proxy.conf
Restarting httpd
Forcing CRL update
CRL generation enabled on the local host. Please make sure to have only a single CRL generation master.
The ipa-crlgen-manage command was successful

Difference in config files seem sane:

--- disabled/CS.cfg	2019-03-12 16:37:23.128831778 +0100
+++ enabled/CS.cfg	2019-03-12 16:37:56.767714341 +0100
@@ -433,2 +433,2 @@
-ca.crl.MasterCRL.enableCRLCache=false
-ca.crl.MasterCRL.enableCRLUpdates=false
+ca.crl.MasterCRL.enableCRLCache=true
+ca.crl.MasterCRL.enableCRLUpdates=true

--- disabled/ipa-pki-proxy.conf	2019-03-12 16:37:23.128831778 +0100
+++ enabled/ipa-pki-proxy.conf	2019-03-12 16:37:56.767714341 +0100
@@ -46,3 +46 @@
-
-
-RewriteRule ^/ipa/crl/MasterCRL.bin http://f29-idm0.laptop.example.org/ca/ee/ca/getCRL?op=getCRL&crlIssuingPoint=MasterCRL [L,R=301,NC]
\ No newline at end of file
+#RewriteRule ^/ipa/crl/MasterCRL.bin http://f29-idm0.laptop.example.org/ca/ee/ca/getCRL?op=getCRL&crlIssuingPoint=MasterCRL [L,R=301,NC]
\ No newline at end of file

When a server is CRL generation master uninstall is prohibited:
[root@f29-idm1 ~]# ipa-crlgen-manage status
CRL generation: enabled
Last CRL update: 2019-03-12 15:44:07
Last CRL Number: 4
The ipa-crlgen-manage command was successful
[root@f29-idm1 ~]# ipa-server-install --uninstall -U
Deleting this server will leave your installation without a CRL generation master.

Aborting uninstall operation.
The ipa-server-install command failed. See /var/log/ipaserver-uninstall.log for more information
[root@f29-idm1 ~]# ipa-crlgen-manage status
CRL generation: enabled
Last CRL update: 2019-03-12 15:44:07
Last CRL Number: 4
The ipa-crlgen-manage command was successful

LGTM.

File: ipaserver/install/ca.py [hunk 2]
```diff
@@ -115,6 +116,39 @@ def print_ca_configuration(options):
...
+    if crlgen_enabled:
+        print("Deleting this server will leave your installation "
+              "without a CRL generation master.")
+        if (options.unattended and not options.ignore_last_of_role) or \
+           not (options.unattended or ipautil.user_input(
+                "Are you sure you want to continue with the uninstall "
+                "procedure?", False)):
+            print("")
+            print("Aborting uninstall operation.")
+            sys.exit(1)
```
Comment (rcritten) [inline ipaserver/install/ca.py]: I believe it should raise a ScriptError() in this case rather than directly exiting.

**Comment by rcritten** [comment]
Otherwise looks good.

File: ipaserver/install/ca.py (hunk 2 above)
Comment (flo-renaud) [reply ipaserver/install/ca.py]: Thanks for the advise, PR updated with your suggestion.

**Comment by rcritten** [comment]
LGTM, remove the temp commit and I'll ack once CI passes.

**Comment by flo-renaud** [comment]
@rcritten 
temp commit removed and CI is now green.

**Comment by flo-renaud** [comment]
master:

* 0d23fa92788b1258005e46920505d9b768d30799 CRL generation master: new utility to enable|disable
* 4e3a64f70316c98a18e403486f6f8afcec5e24e4 Test: add new tests for ipa-crlgen-manage
* 2e73c964e5896eb18006e1ae8b70f2faef484ab3 ipa server: prevent uninstallation if the server is CRL master

**Pull request closed without merging**